### PR TITLE
[Serialization] Add whether allowing errors to the pretty stack output

### PR DESCRIPTION
**Pull request by bnbarham**
It's currently not obvious in crash reports whether compiling with
errors is enabled or not. Since this option can make previously
"impossible" paths now possible, add a message to both the pretty stack
output and fatal deserialization diagnostics to point out that it is
enabled.

Also add some helper methods to check whether the option is enabled or
not - there were getting to be quite a few.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
Ah, I changed `isAllowModuleWithCompilerErrorsEnabled` to check for *both* the module being compiled with allow errors and the option also currently being on. This will now fail in both swift-ide-test and sourcekitd-test as they're not setting it, though they really should. I'll split that off into another PR in the morning, include the cleanup in it, and then have this PR be just adding the pretty stack trace.

File: lib/Serialization/ModuleFile.cpp [hunk 1]
```diff
@@ -117,6 +117,11 @@ ModuleFile::ModuleFile(std::shared_ptr<const ModuleFileSharedCore> core)
   allocateBuffer(Identifiers, core->Identifiers);
 }
 
+bool ModuleFile::isAllowModuleWithCompilerErrorsEnabled() const {
+  return allowedCompilerErrors() &&
+         (!FileContext || getContext().LangOpts.AllowModuleWithCompilerErrors);
```
Comment (xymus) [inline lib/Serialization/ModuleFile.cpp:122]: Could there be a scenario where the reader doesn't accept errors and the module being read did accept errors?

File: lib/FrontendTool/FrontendTool.cpp [hunk 2]
```diff
@@ -2038,6 +2038,11 @@ int swift::performFrontend(ArrayRef<const char *> Args,
     return finishDiagProcessing(1, /*verifierEnabled*/ false);
   }
 
+  Optional<llvm::PrettyStackTraceString> allowErrorsStackTrace;
+  if (Invocation.getFrontendOptions().AllowModuleWithCompilerErrors)
+    allowErrorsStackTrace.emplace("While allowing modules with compiler errors "
+                                  "enabled");
```
Comment (xymus) [inline lib/FrontendTool/FrontendTool.cpp:2044]: Is this check for any frontend jobs and we'd expect the prettystacktrace only for the background indexing then right?

**Review by xymus — APPROVED** [review]
The changes look good and will come in useful I'm sure, thanks!

File: lib/FrontendTool/FrontendTool.cpp (hunk 2 above)
Comment (bnbarham) [reply lib/FrontendTool/FrontendTool.cpp:2044]: Yep, we should only see this message in background indexing processes.

File: lib/Serialization/ModuleFile.cpp (hunk 1 above)
Comment (bnbarham) [reply lib/Serialization/ModuleFile.cpp:122]: There shouldn't be, though that makes me think of the other case - ie. a module that has errors but wasn't compiled with allowing errors. That *should* never happen, but it seems reasonable that we'd just want to continue in that case.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
> Ah, I changed `isAllowModuleWithCompilerErrorsEnabled` to check for _both_ the module being compiled with allow errors and the option also currently being on. This will now fail in both swift-ide-test and sourcekitd-test as they're not setting it, though they really should. I'll split that off into another PR in the morning, include the cleanup in it, and then have this PR be just adding the pretty stack trace.

Ended up setting allow compiler errors in swift-ide-test + passing in that arg to sourcekitd-test in the cleanup commit. I also updated `isAllowModuleWithCompilerErrorsEnabled()` to just `allowCompilerErrors()` and it no longer checks whether the module was compiled with errors - we want to continue regardless.`compiledAllowingCompilerErrors()` is now the method that checks the bit in the module, it's used for the pretty stack print.

**Comment by bnbarham** [comment]
There's a swift-ide-test that relies on allow compiler errors *not* being enabled, so make that an option instead.

@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform

**Review by akyrtzi — APPROVED** [review]

**Comment by bnbarham** [comment]
Added a few tests to actually check for the output + an extra allow errors test case (which seemed weird not to add since the one to check the output assumes that it is valid when allowing errors).

@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test macOS platform

**Comment by bnbarham** [comment]
Added an "asserts" requirement to the crash-adds-message.swift test.

@swift-ci please test

**Pull request merged**